src/agreement/lease-process-pool.ts

```diff
@@ -0,0 +1,285 @@
+  }
+
+  /**
+   * If there are any acquires waiting in the queue, the lease process will be passed to the first one.
+   * Otherwise, the lease process will be added to the queue.
+   */
+  private addLeaseProcessToQueue(leaseProcess: LeaseProcess) {
+    if (this.acquireQueue.length > 0) {
+      const acquire = this.acquireQueue.shift()!;
+      acquire(leaseProcess);
+      return;
+    }
+    if (leaseProcess.hasActivity()) {
+      this.highPriority.add(leaseProcess);
+    } else {
+      this.lowPriority.add(leaseProcess);
+    }
+  }
+
+  async release(leaseProcess: LeaseProcess): Promise<void> {
+    if (this.getAvailable() >= this.maxPoolSize) {
+      return this.destroy(leaseProcess);
+    }
+    this.borrowed.delete(leaseProcess);
+    const isValid = await this.validate(leaseProcess);
+    if (!isValid) {
+      return this.destroy(leaseProcess);
+    }
+    this.events.emit("released", leaseProcess.agreement.getDto());
+    this.addLeaseProcessToQueue(leaseProcess);
+  }
+
+  async destroy(leaseProcess: LeaseProcess): Promise<void> {
+    try {
+      this.borrowed.delete(leaseProcess);
+      this.logger.debug("Destroying lease process from the pool", { agreementId: leaseProcess.agreement.id });
+      await leaseProcess.finalize();
+      this.events.emit("destroyed", leaseProcess.agreement.getDto());
+    } catch (error) {
+      this.events.emit(
+        "error",
+        new GolemMarketError("Destroying lease process failed", MarketErrorCode.LeaseProcessTerminationFailed, error),
+      );
+      this.logger.error("Destroying lease process failed", error);
+    }
+  }
+
+  /**
+   * Sets the pool into draining mode and then clears it
+   *
+   * When set to drain mode, no new acquires will be possible. At the same time, all agreements in the pool will be terminated with the Providers.
+   *
+   * @return Resolves when all agreements are terminated
+   */
+  async drainAndClear() {
+    this.isDraining = true;
+    this.acquireQueue = [];
+    const allLeaseProcesses = Array.from(this.borrowed)
+      .concat(Array.from(this.lowPriority))
+      .concat(Array.from(this.highPriority));
+    await Promise.allSettled(allLeaseProcesses.map((leaseProcess) => this.destroy(leaseProcess)));
+    this.lowPriority.clear();
+    this.highPriority.clear();
+    this.borrowed.clear();
+    this.isDraining = false;
+    this.events.emit("end");
+    return;
+  }
+
+  getSize() {
+    return this.getAvailable() + this.getBorrowed();
+  }
+
+  getAvailable() {
+    return this.lowPriority.size + this.highPriority.size;
+  }
+
+  getBorrowed() {
+    return this.borrowed.size;
+  }
+
+  /**
+   * Wait till the pool is ready to use (min number of items in pool are usable).
+   * Optionally, you can pass an AbortSignal to cancel the operation.
+   */
+  async ready(abortSignal?: AbortSignal): Promise<void> {
+    if (this.minPoolSize <= this.getAvailable()) {
+      return;
+    }
+    const signal = abortSignal || AbortSignal.timeout(10_000);
```

Comment:
`AbortSignal.timeout` -> nice one! I learned something today!
Nevertheless, I don't think that's good to assume this for the user. I would suggest:

- not having this set up by default
- user can opt-in by providing a corresponding option - `readyTimeoutSec` for example

Reply:
1. I think there needs to be some sort of short-circuit when we're dealing with a potentially infinite loop like this one, otherwise this will hang forever (or until ctrl-c)
2. If we use `AbortSignal` then we have both options - a timeout with `AbortSignal.timeout` or manual short-circuit with `AbortController`. Using `readyTimeoutSec: number` gives us only the first option with seemingly no benefit?